### ISSUE TITLE
There is confusing logic in `ddim_sample()` which seems to results in lower performance of multi steps.

### ISSUE THREAD

**Issue by Li-Qingyun** [ShoufaChen/DiffusionDet#56]
Hi, thanks for your work, I have been following DiffusionDet since it was released.

When the sampling timestep is set 4 (default setting of other param, including 500 proposals), I found there are only 1500 boxes in `ensemble` results.
I also tried setting it to 2, there are only 500 proposal and the performance is very close to setting to 1 (but a litter higher, I have not found lower case, more exps require to be done).

It seems that, since the last iteration is stopped by the `continue`, the boxes generated in the last iteration are not appended.

I found the `ddim_sample()` may be based on the [denoising-diffusion-pytorch](https://github.com/lucidrains/denoising-diffusion-pytorch). The `ensemble` operations are similar to the logic controlled by the `return_all_timestep` param [here](https://github.com/lucidrains/denoising-diffusion-pytorch/blob/aab7a9d1faa86e26b8ec3452df773c8d916b154c/denoising_diffusion_pytorch/denoising_diffusion_pytorch.py#L642-L668). I found their implementation about the image appending is also confusing. I created an [issue](https://github.com/lucidrains/denoising-diffusion-pytorch/issues/166) to report and the maintainer modified it in [this commit](https://github.com/lucidrains/denoising-diffusion-pytorch/commit/f4b6a08405c2f02302ce23596b70248601bed41c).

In DiffusionDet, I tried modify `ddim_sample()` like this:
```
if time_next < 0:
    img = x_start
    box_pred_per_image, scores_per_image, labels_per_image = self.inference(outputs_class[-1],
                                                                                        outputs_coord[-1],
                                                                                        images.image_sizes)
    ensemble_score.append(scores_per_image)
    ensemble_label.append(labels_per_image)
    ensemble_coord.append(box_pred_per_image)
    continue
```
It promotes the mAP.

I create this issue to ask whether my opinion is right and help improve the detector.

Moreover, Happy Chinese New Year. 新年快乐！

**Comment by LMD0311** [comment]
Could you tell me the performance gap？ @Li-Qingyun

**Comment by Li-Qingyun** [comment]
> Could you tell me the performance gap？ @Li-Qingyun

It's easy to test. My results are on my own data set.

**Comment by ShoufaChen** [comment]
Hi, @Li-Qingyun 

Good catch. Thanks very much for your detailed findings. I am checking this.

**Comment by LMD0311** [comment]
> > Could you tell me the performance gap？ @Li-Qingyun
> 
> It's easy to test. My results are on my own data set.



I have tried your modification on original code. I tested it on the coco validation set with 2 TITAN Xp GPU, using the weights file provided by the authors. Because of the devices and environment, my results are slightly different from the results in the paper.

|                                    |   AP    | Speed (s / iter) |
| :--------------------------------: | :-----: | :--------------: |
|   Original code with **4** steps   | 45.9163 |     0.198844     |
| Your code with **4** steps | 45.9985 |     0.202367     |
|   Original code with **1** step    | 45.0481 |     0.077163     |

The issues you raise do lead to some performance gains. But overall I feel that `ddim` operation is not cost effective, in other words it's not worth the effort😂

**Comment by Li-Qingyun** [comment]
@ShoufaChen OK
